nixos/modules/services/web-apps/piwik.nix

```diff
@@ -170,11 +167,14 @@ in {
       # References:
       # https://fralef.me/piwik-hardening-with-nginx-and-php-fpm.html
       # https://github.com/perusio/piwik-nginx
-      ${cfg.nginx.virtualHost} = {
-        root = "${pkgs.piwik}/share";
-        enableSSL  = cfg.nginx.enableSSL;
-        enableACME = cfg.nginx.enableACME;
-        forceSSL   = cfg.nginx.forceSSL;
+      "${user}.${config.networking.hostName}" = mkMerge [ cfg.nginx {
+        # don't allow to override root, as it will almost certainly break piwik
+        root = mkForce "${pkgs.piwik}/share";
+
+        # allow to override SSL settings if necessary, i.e. when using another method than ACME
+        # but enable them by default, as sensitive login and piwik data should not be transmitted in clear text.
+        forceSSL = mkDefault true;
```

Comment:
Strangely, the mkDefault options here don't behave like needed: When the user specifies services.piwik.ngnix = {};, this results in adSSL being false, i.e. the default value specified in the original submodule definition at web-servers/nginx/vhost-options.nix. ☹️ On the contrary, root = mkForce… results in the piwik definition.

I don't understand why this happens. Shouldn't a mkOption = { default = false; … }; be equivalent to a mkDefaultOption, and therefore be overwritten by a mkDefault?

By trial and error, I found out the priority has to be less than 100 for the options here to be used when no option is specified. So that's why mkForce works. By searching through the code I now know that 100 is the default priority when no priority is specified. But then of course, if I use a priority < 100, the user can't overwrite it easily anymore.

So still out of ideas how to solve this. ☹️

Reply:
(Was merged a little too early when the original comment went invisible, but as the situation improved from “totally broken” to “broken if you did not explicitly configured SSL but expect it to be default“ and can do that as a workaround for the time being, I guess we can leave it that way. I'll try do deliver a fix in a new PR asap.)

Reply:
Got it! Fix in #28795 , key was using `recursiveUpdate`.